doc/language_reference/language_reference.md

```diff
@@ -242,9 +243,18 @@ typedef R = R{f1: bigint, f2:bool}
 relation R[R]
 ```
 
+The optional primary key clause is only allowed in `input` relations and defines a mapping from a record to
+its unique key. For example, the following declares a relation with primary key `f1`:
+
+```
+relation R(f1: bigint, f2: bool) 
+primary key (r) r.f1
```

Comment:
Relations don't really have fields. A relation can have any type, including scalar type without fields like `bool`, a tuple that also does not have named fields or a variant type with multiple constructors. And even in the common case where relation is just a struct with fields, it might be necessary to access the record as a whole, e.g., to pass it as an argument to a function. This is why I needed a way to refer to the entire record by name. This is essentially a lambda, except that we don't support lambdas right now.

Here are some options:
- we can come up with a shortcut syntax for the common case when the primary key is just a subset of fields, e.g., `primary key (f1, f2)`
- we can come up with a more general syntax for lambda expressions and use it here, so that we don't need to change the primary key syntax when we add support for lambdas

Reply:
For now you can leave it as it is, but we should do both these things.
For lambdas I would suggest `(r) -> r.f1`. Shouldn't be much harder to parse.